Sometimes the CI will fail due to a transport error

The CI will ocassionally fail due to this error:

```
transport_test.go:240: error upserting: unable to commit transaction: commit transaction: (WriteConflict) WriteConflict error: this operation conflicted with another operation. Please retry your operation or multi-document transaction.
```

It's likely that a test is conflicting with an operation in a test in the storage package, all of these tests run in parallel.

--- a/internal/storage/mongo.go
+++ b/internal/storage/mongo.go
@@ -10,6 +10,7 @@ package storage
 import (
 	"context"
 	"fmt"
+	"sync"
 	"time"
 
 	"github.com/alpine-hodler/gidari/proto"
@@ -26,8 +27,9 @@ const defaultMDBLifetime = 60 * time.Second
 // Mongo is a wrapper for *mongo.Client, use to perform CRUD operations on a mongo DB instance.
 type Mongo struct {
 	*mongo.Client
-	dns      string
-	lifetime time.Duration
+	dns        string
+	lifetime   time.Duration
+	writeMutex sync.Mutex
 }
 
 // NewMongo will return a new mongo client that can be used to perform CRUD operations on a mongo DB instance. This
@@ -45,6 +47,7 @@ func NewMongo(ctx context.Context, uri string) (*Mongo, error) {
 	mdb.Client = client
 	mdb.dns = uri
 	mdb.lifetime = defaultMDBLifetime
+	mdb.writeMutex = sync.Mutex{}
 
 	return mdb, nil
 }
@@ -177,6 +180,9 @@ func (m *Mongo) Truncate(ctx context.Context, req *proto.TruncateRequest) (*prot
 
 // Upsert will insert or update a record in a collection.
 func (m *Mongo) Upsert(ctx context.Context, req *proto.UpsertRequest) (*proto.UpsertResponse, error) {
+	m.writeMutex.Lock()
+	defer m.writeMutex.Unlock()
+
 	records, err := tools.DecodeUpsertRecords(req)
 	if err != nil {
 		return nil, fmt.Errorf("failed to decode records: %w", err)

--- a/internal/storage/postgres.go
+++ b/internal/storage/postgres.go
@@ -263,6 +263,9 @@ func (pg *Postgres) getPrepareContextFn(ctx context.Context) (sqlPrepareContextF
 // PK on the request record to update the data in the database. An upsert request will update the entire table
 // for a given record, include fields that have not been set directly.
 func (pg *Postgres) Upsert(ctx context.Context, req *proto.UpsertRequest) (*proto.UpsertResponse, error) {
+	pg.writeMutext.Lock()
+	defer pg.writeMutext.Unlock()
+
 	records, err := tools.DecodeUpsertRecords(req)
 	if err != nil {
 		return nil, fmt.Errorf("unable to decode records: %w", err)
@@ -307,8 +310,10 @@ type Postgres struct {
 	*sql.DB
 
 	// meta hold metdata about the database.
-	meta      *pgmeta
-	metaMutex sync.Mutex
+	meta *pgmeta
+
+	metaMutex   sync.Mutex
+	writeMutext sync.Mutex
 
 	// activeTx are the transactions that are currently active on this connection. When a user calls "StartTx" on
 	// a Postgres intance, a transaction is created and added to this map. Afterward, if the user calls a write
@@ -333,6 +338,7 @@ func NewPostgres(ctx context.Context, connectionURL string) (*Postgres, error) {
 	postgres.setMaxOpenConns()
 	postgres.meta = new(pgmeta)
 	postgres.metaMutex = sync.Mutex{}
+	postgres.writeMutext = sync.Mutex{}
 	postgres.activeTx = sync.Map{}
 
 	return postgres, nil

